File: lib/services/kem.dart
import 'dart:typed_data';

import '../src/rust/api/kem.dart' as rust;

/// KEM post-cuántico (libcrux): X25519, P256, ML-KEM 512/768/1024
/// e híbridos X25519+ML-KEM768 / X-Wing.
///
/// Flujo: A genera par → B encapsula contra la pública de A → B manda el
/// ciphertext → A desencapsula → ambos tienen el MISMO shared secret.
class Kem {
  Future<List<String>> listAlgorithms() async => rust.kemListAlgorithms();

  Future<KemKeyPair> keyGen(String algorithm) async {
    final kp = await rust.kemKeyGen(algorithm: algorithm);
    return KemKeyPair(
      privateKey: Uint8List.fromList(kp.privateKey),
      publicKey: Uint8List.fromList(kp.publicKey),
    );
  }

  Future<KemEncapsulation> encapsulate({
    required String algorithm,
    required Uint8List publicKey,
  }) async {
    final e = await rust.kemEncapsulate(
        algorithm: algorithm, publicKey: publicKey);
    return KemEncapsulation(
      sharedSecret: Uint8List.fromList(e.sharedSecret),
      ciphertext: Uint8List.fromList(e.ciphertext),
    );
  }

  Future<Uint8List> decapsulate({
    required String algorithm,
    required Uint8List ciphertext,
    required Uint8List privateKey,
  }) async {
    return Uint8List.fromList(await rust.kemDecapsulate(
      algorithm: algorithm,
      ciphertext: ciphertext,
      privateKey: privateKey,
    ));
  }
}

class KemKeyPair {
  final Uint8List privateKey;
  final Uint8List publicKey;
  KemKeyPair({required this.privateKey, required this.publicKey});
}

class KemEncapsulation {
  final Uint8List sharedSecret;
  final Uint8List ciphertext;
  KemEncapsulation({required this.sharedSecret, required this.ciphertext});
}

File: rust/src/api/kem.rs
/// KEM post-cuántico (libcrux) — portado de Gtool `kem_godot.rs` sin Godot.
///
/// Algoritmos: X25519, Secp256r1, ML-KEM 512/768/1024 (FIPS 203) e
/// híbridos X25519+ML-KEM768 / X-Wing.
use libcrux_kem::{self, Algorithm, Ct, PrivateKey, PublicKey};

fn parse_algorithm(name: &str) -> Option<Algorithm> {
    match name {
        "X25519" => Some(Algorithm::X25519),
        "Secp256r1" => Some(Algorithm::Secp256r1),
        "MlKem512" => Some(Algorithm::MlKem512),
        "MlKem768" => Some(Algorithm::MlKem768),
        "MlKem1024" => Some(Algorithm::MlKem1024),
        "X25519MlKem768Draft00" => Some(Algorithm::X25519MlKem768Draft00),
        "XWingKemDraft06" => Some(Algorithm::XWingKemDraft06),
        _ => None,
    }
}

/// Lista de algoritmos disponibles (para llenar un dropdown).
#[flutter_rust_bridge::frb]
pub fn kem_list_algorithms() -> Vec<String> {
    vec![
        "X25519".into(),
        "Secp256r1".into(),
        "MlKem512".into(),
        "MlKem768".into(),
        "MlKem1024".into(),
        "X25519MlKem768Draft00".into(),
        "XWingKemDraft06".into(),
    ]
}

/// Par de claves KEM codificadas.
#[flutter_rust_bridge::frb]
pub struct KemKeyPair {
    pub private_key: Vec<u8>,
    pub public_key: Vec<u8>,
}

/// Resultado de encapsulación.
#[flutter_rust_bridge::frb]
pub struct KemEncapsulation {
    pub shared_secret: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

/// Genera par de claves para el algoritmo dado.
#[flutter_rust_bridge::frb]
pub fn kem_key_gen(algorithm: String) -> Result<KemKeyPair, String> {
    let alg = parse_algorithm(&algorithm)
        .ok_or_else(|| format!("algoritmo desconocido: {algorithm}"))?;

    use rand::rngs::OsRng;
    use rand::TryRngCore;
    let mut os_rng = OsRng;
    let mut rng = os_rng.unwrap_mut();

    let (sk, pk) =
        libcrux_kem::key_gen(alg, &mut rng).map_err(|e| format!("key_gen error: {e:?}"))?;
    Ok(KemKeyPair {
        private_key: sk.encode(),
        public_key: pk.encode(),
    })
}

/// Encapsula contra una clave pública: shared secret + ciphertext.
#[flutter_rust_bridge::frb]
pub fn kem_encapsulate(
    algorithm: String,
    public_key: Vec<u8>,
) -> Result<KemEncapsulation, String> {
    let alg = parse_algorithm(&algorithm)
        .ok_or_else(|| format!("algoritmo desconocido: {algorithm}"))?;
    let pk = PublicKey::decode(alg, &public_key)
        .map_err(|e| format!("error decodificando public key: {e:?}"))?;

    use rand::rngs::OsRng;
    use rand::TryRngCore;
    let mut os_rng = OsRng;
    let mut rng = os_rng.unwrap_mut();

    let (ss, ct) = pk.encapsulate(&mut rng).map_err(|e| format!("encapsulate error: {e:?}"))?;
    Ok(KemEncapsulation {
        shared_secret: ss.encode(),
        ciphertext: ct.encode(),
    })
}

/// Desencapsula con la clave privada → mismo shared secret.
#[flutter_rust_bridge::frb]
pub fn kem_decapsulate(
    algorithm: String,
    ciphertext: Vec<u8>,
    private_key: Vec<u8>,
) -> Result<Vec<u8>, String> {
    let alg = parse_algorithm(&algorithm)
        .ok_or_else(|| format!("algoritmo desconocido: {algorithm}"))?;
    let ct =
        Ct::decode(alg, &ciphertext).map_err(|e| format!("error decodificando ciphertext: {e:?}"))?;
    let sk = PrivateKey::decode(alg, &private_key)
        .map_err(|e| format!("error decodificando private key: {e:?}"))?;
    let ss = ct.decapsulate(&sk).map_err(|e| format!("decapsulate error: {e:?}"))?;
    Ok(ss.encode())
}
